drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright (c) 2020, Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT quicklogic_eos_s3_gpio
+
+#include <errno.h>
+#include <drivers/gpio.h>
+#include <soc.h>
+#include <eoss3_hal_gpio.h>
+#include <eoss3_hal_pads.h>
+#include <eoss3_hal_pad_config.h>
+
+#include "gpio_utils.h"
+
+#define MAX_GPIOS_PER_PORT	8
+#define PORTS_NUM		2
+#define MAX_GPIOS		(MAX_GPIOS_PER_PORT * PORTS_NUM)
+
+struct gpio_eos_s3_config {
+};
+
+struct gpio_eos_s3_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	/* port ISR callback routine address */
+	sys_slist_t callbacks;
+	/* Runtime configuration */
+	uint8_t gpio_irq_num[MAX_GPIOS];
+};
+
+/*
+ * The PAD (pin) value can be greater than 31 in EOS-S3,
+ * because of that it is assumed that instead of the PAD number,
+ * the GPIO number is passed as the `gpio_pin_t` argument in `pin_configure`
+ * call to avoid errors from the high level API which uses BIT()
+ * macro to point the bit corresponding to target pin.
+ *
+ * Additionally, there are only 8 GPIOs in EOS-S3 and there are always 2 PADs
+ * which can be configured to each GPIO. Since each GPIO has two possibilities
+ * of connecting to the one of PADs, there are two mockup ports
+ * defined to distinguish these configurations. Because each port has
+ * the same IRQn we cannot define separate devices in the device tree.
+ * In this case to configure either "port0" or "port1" it is assmued that GPIOs
+ * in range of 0 to 7 point to "port0" and GPIOs indexed from 8 to 15 point
+ * to "port1".
+ */
+static int find_pad(GPIOCfgTypeDef *cfg, int port_num)
+{
```

Comment:
Can you use the arguments to index a static const array of 16 `PadConfig` instances?  I suspect the cost in flash will be significantly less than the size of this function that constructs them at runtime.

    cfg->xPadConf = pad_array[(8 * port_num) + cfg->ucGpioNum];